src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -1009,19 +1020,19 @@ export class SavedObjectsRepository {
     if (!this._allowedTypes.includes(type)) {
       throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
     }
-
     const namespace = normalizeNamespace(options.namespace);
-
-    const { body, statusCode } = await this.client.get<SavedObjectsRawDocSource>(
+    const { body, statusCode, headers } = await this.client.get<SavedObjectsRawDocSource>(
       {
         id: this._serializer.generateRawId(namespace, type, id),
         index: this.getIndexForType(type),
       },
       { ignore: [404] }
     );
-
     const indexNotFound = statusCode === 404;
-
+    // check if we have the elasticsearch header when doc.found is not true (false, undefined or null) and if we do, ensure it is Elasticsearch
+    if (!isFoundGetResponse(body) && !isEsHeaderValid(headers)) {
```

Comment:
If we decide to use the same `isEsUnavailable` error (see comment above), then I think we should simplify this condition to only `if (!isEsHeaderValid(headers))` to catch this issue more broadly. Same for the other repository functions.

Reply:
`isEsUnavailable` does not return an error. It returns a `boolean`.

Reply:
In addition, what we're hoping to accomplish with this work is to differentiate between a saved object not actually being found (404) vs the response returning a 404 because ES is unavailable. For the initial work (this PR), we're only scoping the work to checking if the response is from ES _when we get a 404_, rather than a larger check on _all_ responses.

We will need to implement something for ensuring all responses received are valid ES responses in the long run but IMO, is too risky to try and get in for a patch release.